src/plugins/ucc_discovery_plugin.cpp

```diff
@@ -0,0 +1,368 @@
+      }
+
+      if (type == LQPNodeType::Aggregate) {
+        _ucc_candidates_from_aggregate_node(node, ucc_candidates);
+        return LQPVisitation::VisitInputs;
+      }
+      // no need for if statement; only remaining option is a join node
+      _ucc_candidates_from_join_node(node, ucc_candidates);
+
+      return LQPVisitation::VisitInputs;
+    });
+  }
+
+  return ucc_candidates;
+}
+
+void UccDiscoveryPlugin::_validate_ucc_candidates(const UCCCandidates& ucc_candidates) const {
+  for (const auto& candidate : ucc_candidates) {
+    auto candidate_time = Timer();
+    const auto table = Hyrise::get().storage_manager.get_table(candidate.table_name);
+    const auto col_id = candidate.column_id;
+
+    std::ostringstream message;
+
+    message << "Checking candidate " << candidate.table_name << "." << table->column_name(col_id);
+
+    const auto& soft_key_constraints = table->soft_key_constraints();
+
+    // Skip already discovered unique constraints.
+    if (std::any_of(soft_key_constraints.cbegin(), soft_key_constraints.cend(), [&col_id](const auto& key_constraint) {
+          const auto& columns = key_constraint.columns();
+          return columns.size() == 1 && columns.contains(col_id);
+        })) {
+      message << " [skipped (already known) in " << format_duration(candidate_time.lap()) << "]";
+      Hyrise::get().log_manager.add_message("UccDiscoveryPlugin", message.str(), LogLevel::Info);
+      continue;
+    }
+
+    resolve_data_type(table->column_data_type(col_id), [&](const auto data_type_t) {
+      using ColumnDataType = typename decltype(data_type_t)::type;
+
+      // utilize efficient check for uniqueness inside each dictionary segment for a potential early out
+      if (!_uniqueness_holds_in_dictionary_segments<ColumnDataType>(table, col_id)) {
+        message << " [rejected in " << format_duration(candidate_time.lap()) << "]";
+        Hyrise::get().log_manager.add_message("UccDiscoveryPlugin", message.str(), LogLevel::Info);
+        return;
+      }
+
+      // If we reach here, we have to run the more expensive cross-segment duplicate check
+      if (!_uniqueness_holds_across_segments<ColumnDataType>(table, col_id)) {
+        message << " [rejected in " << format_duration(candidate_time.lap()) << "]";
+        Hyrise::get().log_manager.add_message("UccDiscoveryPlugin", message.str(), LogLevel::Info);
+        return;
+      }
+
+      // We save UCC constraints directly inside the table so they can be forwarded to nodes in a query plan.
+      message << " [confirmed in " << format_duration(candidate_time.lap()) << "]";
+      Hyrise::get().log_manager.add_message("UccDiscoveryPlugin", message.str(), LogLevel::Info);
+      table->add_soft_key_constraint(TableKeyConstraint({col_id}, KeyConstraintType::UNIQUE));
+    });
+  }
+  Hyrise::get().log_manager.add_message("UccDiscoveryPlugin", "Clearing LQP and PQP cache...", LogLevel::Debug);
+
+  Hyrise::get().default_lqp_cache->clear();
```

Comment:
Why clean? I am curious as the plan cache is the single source of information for multiple "autonomous components".

Reply:
Without cleaning or replacing the cache, optimized versions using UCCs will not be applied when a query is executed before and after discovery.

Reply:
Makes sense. But I don't like the answer.  :D